flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchAsyncWriter.java

```diff
@@ -0,0 +1,308 @@
+                "Opensearch AsyncWwriter failed to persist {} entries (retryable = {})",
+                requestEntries.size(),
+                retryable,
+                err);
+
+        numRecordsOutErrorsCounter.inc(requestEntries.size());
+
+        if (retryable) {
+            requestResult.accept(requestEntries);
+        }
+    }
+
+    private void handlePartiallyFailedBulkRequests(
+            BulkResponse response,
+            List<DocSerdeRequest<?>> requestEntries,
+            Consumer<List<DocSerdeRequest<?>>> requestResult) {
+
+        final List<DocSerdeRequest<?>> failedRequestEntries = new ArrayList<>();
+        final BulkItemResponse[] items = response.getItems();
+
+        for (int i = 0; i < items.length; i++) {
+            if (items[i].getFailure() != null) {
+                failedRequestEntries.add(DocSerdeRequest.from(requestEntries.get(i).getRequest()));
+            }
+        }
+
+        numRecordsOutErrorsCounter.inc(failedRequestEntries.size());
+        requestResult.accept(failedRequestEntries);
+    }
+
+    private static RestClientBuilder configureRestClientBuilder(
+            RestClientBuilder builder, NetworkClientConfig networkClientConfig) {
+        if (networkClientConfig.getConnectionPathPrefix() != null) {
+            builder.setPathPrefix(networkClientConfig.getConnectionPathPrefix());
+        }
+
+        builder.setHttpClientConfigCallback(
+                httpClientBuilder -> {
+                    if (networkClientConfig.getPassword() != null
+                            && networkClientConfig.getUsername() != null) {
+                        final CredentialsProvider credentialsProvider =
+                                new BasicCredentialsProvider();
+                        credentialsProvider.setCredentials(
+                                AuthScope.ANY,
+                                new UsernamePasswordCredentials(
+                                        networkClientConfig.getUsername(),
+                                        networkClientConfig.getPassword()));
+
+                        httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+                    }
+
+                    if (networkClientConfig.isAllowInsecure().orElse(false)) {
+                        try {
+                            httpClientBuilder.setSSLContext(
+                                    SSLContexts.custom()
+                                            .loadTrustMaterial(new TrustAllStrategy())
+                                            .build());
+                        } catch (final NoSuchAlgorithmException
+                                | KeyStoreException
+                                | KeyManagementException ex) {
+                            throw new IllegalStateException(
+                                    "Unable to create custom SSL context", ex);
+                        }
+                    }
+
+                    return httpClientBuilder;
```

Comment:
nit: Should we move this out to a separate class?

Reply:
why? that's the only place is needed actually, seems like sealing it in place is acceptable

Reply:
In my opinion it breaches the single responsibility philosophy. The writer is responsible for writing and knowing how to construct the client. I am less concerned how many times it is used. However, I marked as _nit_ since I am not marking this as a blocker.

Reply:
Thanks @dannycranmer , I think this is a good idea (more over, I was not correct, there was another place with the similar instantiation logic present), extracted the utility class